### PR TITLE
feat: treat empty string as none when formating

### LINKED ISSUE #1855: Conditional format string in prompt always renders despite containing empty variables
#### Current Behavior
The final part of my prompt always renders, despite being a conditional format string that always contains empty variables unless in a git directory.

#### Expected Behavior
The final part of my prompt should only render when I am in a git directory, [as stated in the documentation](https://starship.rs/config/#conditional-format-strings).

#### Additional context/Screenshots
In the following screenshot, you can see how the prompt looks both when in and out of a git directory. In the first line, the last arrow should not be rendered because it is part of a conditional format string where all the variables are empty.
prompt_problem

#### Environment
- Starship version: 0.46.2
- zsh version: zsh 5.8 (x86_64-unknown-linux-gnu)
- Operating system: Linux Rolling Release
- Terminal emulator: alacritty 0.5.0

#### Relevant Shell Configuration

```bash
# User configuration sourced by interactive shells

# Remove older command from the history if a duplicate is to be added.
setopt HIST_IGNORE_ALL_DUPS

# Remove path separator from WORDCHARS.
WORDCHARS=${WORDCHARS//[\/]}

# Lines configured by zsh-newuser-install
HISTFILE=~/.histfile
HISTSIZE=1000
SAVEHIST=1000
bindkey -v
# End of lines configured by zsh-newuser-install
# The following lines were added by compinstall
zstyle :compinstall filename '/home/erebus/.zshrc'

autoload -Uz compinit
compinit
# End of lines added by compinstall

# antigen
source $HOME/bin/antigen.zsh
antigen bundle zsh-users/zsh-completions
antigen bundle zsh-users/zsh-history-substring-search
antigen bundle zsh-users/zsh-syntax-highlighting
antigen apply
bindkey '^[[A' history-substring-search-up
bindkey '^[[B' history-substring-search-down
bindkey -M vicmd 'k' history-substring-search-up
bindkey -M vicmd 'j' history-substring-search-down

# exports for sessions
export LANG=en_US.UTF-8
export LC_CTYPE=en_US.UTF-8
export EDITOR=nano
export TUIR_EDITOR=nano
export TUIR_BROWSER=firefox
export TUIR_URLVIEWER=urlview
export CLIPPY_DISABLE_DOCS_LINKS=true
export RUST_SRC_PATH="$(rustc --print sysroot)/lib/rustlib/src/rust/src"

# useful aliases
alias sudo='sudo '
alias doas='doas '
alias vpm='vpm --color=yes'
alias g='git'
alias ls='ls --color=always --group-directories-first'
alias lg='lazygit'
alias j='just'
alias tb='nc termbin.com 9999'

# gpg configuration
export GPG_TTY="$(tty)"
export SSH_AUTH_SOCK=$(gpgconf --list-dirs agent-ssh-socket)
gpgconf --launch gpg-agent

# prompt
eval "$(starship init zsh)"
```

#### Starship Configuration

```toml
add_newline = false

format = """
$character\
$jobs\
${custom.plan9}\
[❯ ](bold white)\
$directory\
[❯ ](bold white)\
($git_branch\
$git_status\
$git_state\
[❯ ](bold white))"""

[aws]
disabled = true

[battery]
disabled = true

[character]
success_symbol = "[λ](bold green)"
error_symbol = "[λ](bold red)"
disabled = false

[cmake]
disabled = true

[cmd_duration]
disabled = true

[conda]
disabled = true

[crystal]
disabled = true

[dart]
disabled = true

[directory]
truncation_length = 1
fish_style_pwd_dir_length = 1
disabled = false

[docker_context]
disabled = true

[dotnet]
disabled = true

[elixir]
disabled = true

[elm]
disabled = true

[env_var]
disabled = true

[erlang]
disabled = true

[gcloud]
disabled = true

[git_branch]
symbol = " "
format = "[$symbol$branch]($style) "
disabled = false

[git_commit]
disabled = true

[git_state]
disabled = false

[git_status]
disabled = false

[golang]
disabled = true

[helm]
disabled = true

[hostname]
disabled = true

[java]
disabled = true

[jobs]
threshold = 0
symbol = "+"
style = "cyan"
disabled = false

[julia]
disabled = true

[kubernetes]
disabled = true

[line_break]
disabled = true

[lua]
disabled = true

[memory_usage]
disabled = true

[hg_branch]
disabled = true

[nim]
disabled = true

[nix_shell]
disabled = true

[nodejs]
disabled = true

[package]
disabled = true

[ocaml]
disabled = true

[openstack]
disabled = true

[perl]
disabled = true

[php]
disabled = true

[purescript]
disabled = true

[python]
disabled = true

[ruby]
disabled = true

[rust]
disabled = true

[shlvl]
disabled = true

[singularity]
disabled = true

[swift]
disabled = true

[status]
disabled = true

[terraform]
disabled = true

[time]
disabled = true

[username]
disabled = true

[zig]
disabled = true

[custom.plan9]
command = "echo ❾"
when = """ echo $PATH | grep "^/usr/lib/plan9/bin" """
description = "Indicates that the user is in a plan9-like environment"
style = "purple"
disabled = false

```

### LINKED ISSUE #2700: treat empty string as none when formating
## Feature Request

currently, only the disabled module can disable a conditional format group(I didn't know how to call it, But I'm saying the syntax get tested here https://github.com/starship/starship/blob/master/src/formatter/string_formatter.rs#L565 ).

But something a module is not working will return an empty string but not none, which can't disable this conditional formating string.

this is useful when I want to wrap all git status in to brackets like `(master+)` or `(v1.0.0 2k2ksd8)` with `(\($git_branch$git_commit$git_state$git_status\))`

#### Is your feature request related to a problem? Please describe.

No

#### Describe the solution you'd like

```TOML
format = "1(2$aws)3"
[aws]
disabled = false
format = ''
```
get rendered to `13` instead of current `123`

### PR DESCRIPTION
**Pull request by cymruu**
String formatter treats empty string as it was `None`

#### Motivation and Context
Closes #2700
Closes #1855

#### How Has This Been Tested?
- [x] I have tested using **MacOS**
- [x] I have tested using **Linux**
- [ ] I have tested using **Windows**

#### Checklist:

- [x] I have updated the documentation accordingly.
- [x] I have updated the tests accordingly.

### CODE REVIEW

File: docs/config/README.md
Comment (andytom) [inline docs/config/README.md:112]: Should this also be updated to the following? 
```suggestion
 - When `$all` is a shortcut for `\[$a$b\] `, `($all)` will show nothing only if `$a` and `$b` are both `None` or empty strings.
```

**Review by andytom — APPROVED** [review]
Have left a suggestion but other than that LGTM. Note that I have not tried this out yet so would be good to get some more eyes on this before merging.

File: docs/config/README.md
Comment (cymruu) [reply docs/config/README.md:112]: I've create such test and it fails so I think no. Should it also handle that case? I don't quite understand `map_meta` for now
```rust
    #[test]
    fn test_empty_meta_variable() {
        const FORMAT_STR: &str = "$all";
        const FORMAT_STR__ALL: &str = "$a$b";

        let formatter = StringFormatter::new(FORMAT_STR)
            .unwrap()
            .map_meta(|var, _| match var {
                "all" => Some(FORMAT_STR__ALL),
                _ => None,
            })
            .map(|var| match var {
                "a" => Some(Ok("")),
                "b" => Some(Ok("")),
                _ => None,
            });
        let result = formatter.parse(None).unwrap();
        assert_eq!(result.len(), 0);
    }
```

Comment (andytom) [reply docs/config/README.md:112]: I think we should probably support this but maybe we can do it later in a separate PR.

**Comment by cinerea0** [comment]
Are there any remaining roadblocks to this being merged?

**Review by matchai — APPROVED** [review]
Looks good to me and works excellently in testing! 👍 
Thank you for your contribution, Filip! 🎉

**Pull request merged**
